hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2017,3 +2032,27 @@ func (r *NodePoolReconciler) doesCPODecompressAndDecodeConfig(ctx context.Contex
 	_, managesDecompressAndDecodeConfig := supportutil.ImageLabels(controlPlaneOperatorImageMetadata)[controlPlaneOperatorManagesDecompressAndDecodeConfig]
 	return managesDecompressAndDecodeConfig, nil
 }
+
+// ensureMachineDeletion ensures all the machines belonging to the NodePool's MachineSet are fully deleted
+// This function can be deleted once the upstream PR (https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3805) is merged and pulled into https://github.com/openshift/cluster-api
+// This function is necessary to ensure AWSMachines are fully deleted prior to deleting the NodePull secrets being deleted due to a bug introduced by https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271
+// See https://github.com/openshift/hypershift/pull/1826#discussion_r1007349564 for more details
+func (r *NodePoolReconciler) ensureMachineDeletion(nodePool *hyperv1.NodePool, controlPlaneNamespace string) error {
+	// Get list of CAPI Machines to filter through
+	machines := capiv1.MachineList{}
+	if err := r.List(context.Background(), &machines, &client.ListOptions{Namespace: controlPlaneNamespace}); err != nil {
+		return fmt.Errorf("failed to get list of Machines: %w", err)
+	}
+
+	// Filter out only machines belonging to deleted NodePool
+	var machineSetOwnedMachines []capiv1.Machine
+	for i, machine := range machines.Items {
+		if machine.Annotations[nodePoolAnnotation] == client.ObjectKeyFromObject(nodePool).String() {
+			machineSetOwnedMachines = append(machineSetOwnedMachines, machines.Items[i])
+		}
+	}
+	if len(machineSetOwnedMachines) > 0 {
+		return fmt.Errorf("")
```

Comment:
add message to return e.g. fmt.Errorf("there are still Machines in the cluster")

Reply:
Ok, it's been updated